diffrax/integrate.py

```diff
@@ -142,34 +218,107 @@ def body_fun(state, inplace):
+                        if is_vf_wrapper(x):
+                            vector_field = _HistoryVectorField(
+                                x.vector_field,
+                                t0,
+                                state.tprev,
+                                state.tnext,
+                                dense_info,
+                                dense_interp,
+                                solver.interpolation_cls,
+                                y0_history,
+                                delays.delays,
+                            )
+                            return VectorFieldWrapper(vector_field)
+                        else:
+                            return x
+
+                    terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
+                    (y, y_error, dense_info, solver_state, solver_result) = solver.step(
+                        terms_,
+                        state.tprev,
+                        state.tnext,
+                        state.y,
+                        args,
+                        state.solver_state,
+                        state.made_jump,
+                    )
 
-            terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
-            # TODO: write down implicit problem wrt dense_info, using `terms_`
-            (y, y_error, dense_info, solver_state, solver_result) = terms_  # ...
+                    _pred = (((y - y_prev) / y) > delays.eps).any()
+                    _pred = _pred & (step < 10)
+                    return (
+                        y,
+                        y_error,
+                        dense_info,
+                        solver_state,
+                        solver_result,
+                        _pred,
+                        step + 1,
+                    )
+
+                _init_val = (
+                    state.y,
+                    0,
+                    dict(
+                        zip(
+                            state.dense_infos.keys(),
+                            [
+                                state.dense_infos[key][state.dense_save_index - 1]
+                                for key in state.dense_infos.keys()
+                            ],
+                        )
+                    ),
+                    0,
+                    0,
+                    True,
+                    0,
+                )
+                (
+                    y,
+                    y_error,
+                    dense_info,
+                    solver_state,
+                    solver_result,
+                    _,
+                    _,
+                ) = lax.while_loop(_cond_fun, _body_fun, _init_val)
+                return y, y_error, dense_info, solver_state, solver_result
+
+            (
+                y,
+                y_error,
+                dense_info,
+                solver_state,
+                solver_result,
+            ) = history_extrapolation_implicit()
```

Comment:
We discussed moving this new logic (and other related quantities like `Delays`) into a new file. I think now is the time to do that, as I'm broadly happy with how this looks now.

Reply:
(If it ends up being cleaner to review the diffs, then you could move the target branch from `patrick-kidger:delay` to `patrick-kidger:main`, so we don't see the deletion of my demonstration code in the diff.)

Reply:
Sound goods ill do in next commit